dspace-api/src/main/java/org/dspace/harvest/dao/impl/HarvestedCollectionDAOImpl.java

```diff
@@ -39,98 +40,136 @@ protected HarvestedCollectionDAOImpl() {
+        hasCorrectStatusOrIsExpiredRestrictions.add(harvestExpiredRestriction);
 
-            )
+        Predicate hasCorrectStatusOrIsExpiredPredicate = criteriaBuilder.or(hasCorrectStatusOrIsExpiredRestrictions
+                                                                                .toArray(new Predicate[] {}));
+
+        Predicate hasMinimalType = criteriaBuilder.greaterThan(
+            harvestedCollectionRoot.get(HarvestedCollection_.harvestType),
+            minimalType);
+
+        criteriaQuery.where(criteriaBuilder.and(wasNotHarvestedInCurrentRun,
+                                                hasMinimalType,
+                                                hasCorrectStatusOrIsExpiredPredicate
+                            )
         );
-        criteria.addOrder(Order.asc("lastHarvested"));
-        return list(criteria);
+
+        List<javax.persistence.criteria.Order> orderList = new LinkedList<>();
+        orderList.add(criteriaBuilder.asc(harvestedCollectionRoot.get(HarvestedCollection_.lastHarvested)));
+        criteriaQuery.orderBy(orderList);
+
+
+        return list(context, criteriaQuery, false, HarvestedCollection.class, -1, -1);
+
 
     }
 
     @Override
     public int count(Context context) throws SQLException {
-        Criteria criteria = createCriteria(context, HarvestedCollection.class);
-        return count(criteria);
+
+        CriteriaBuilder criteriaBuilder = getCriteriaBuilder(context);
+        CriteriaQuery<Long> criteriaQuery = criteriaBuilder.createQuery(Long.class);
+
+        Root<HarvestedCollection> harvestedCollectionRoot = criteriaQuery.from(HarvestedCollection.class);
+        return count(context, criteriaQuery, criteriaBuilder, harvestedCollectionRoot);
     }
 
-    protected Criteria getByStatusAndMinimalTypeCriteria(Context context, int status, int type, int limit)
+    protected CriteriaQuery getByStatusAndMinimalTypeCriteria(Context context, int status, int type, int limit)
         throws SQLException {
-        Criteria criteria = createCriteria(context, HarvestedCollection.class);
-        criteria.add(
-            Restrictions.and(
-                Restrictions.gt("harvestType", type),
-                Restrictions.eq("harvestStatus", status)
-            )
+        CriteriaBuilder criteriaBuilder = getCriteriaBuilder(context);
+        CriteriaQuery criteriaQuery = getCriteriaQuery(criteriaBuilder, HarvestedCollection.class);
+        Root<HarvestedCollection> harvestedCollectionRoot = criteriaQuery.from(HarvestedCollection.class);
+        criteriaQuery.select(harvestedCollectionRoot);
+        criteriaQuery.where(criteriaBuilder.and(
+            criteriaBuilder.greaterThan(harvestedCollectionRoot.get(HarvestedCollection_.harvestType), type),
+            criteriaBuilder.equal(harvestedCollectionRoot.get(HarvestedCollection_.harvestStatus), status)
+                            )
         );
-        if (limit != -1) {
-            criteria.setMaxResults(1);
```

Comment:
We seem to have lost 'limit'.

Reply:
The way that this method is currently structured by returning a criteriaQuery implies that this method should not set a limit on the results, that has to be done later down the line when a `Query` is made from the `CriteriaQuery` object that this method will return. 
However this method is never used in the current DSpace code. I'll remove the limit parameter and leave the method as is unless we want to return a list of HarvestCollections here instead of the CriteriaQuery, in which case we can call the logic to execute the query and pass along the limit.